p2p/discovery.py

```diff
@@ -820,6 +946,63 @@ def __init__(self, proto: DiscoveryProtocol, peer_pool: PeerPool,
         await self.proto.stop()
 
 
+class NodeTicketInfo:
+    last_issued: int = 0
+    last_used: int = 0
+
+
+class TopicEntry:
+
+    def __init__(self, node: kademlia.Node, expiry: float) -> None:
+        self.node = node
+        self.expiry = expiry
+
+
+class TopicTable:
+    registration_lifetime = 60 * 60
+
+    def __init__(self, logger: TraceLogger) -> None:
+        self.logger = logger
+        # A per-topic FIFO list of nodes.
+        self.topics: Dict[bytes, List[TopicEntry]] = {}
+        # The IDs of the last issued/used tickets for any given node.
+        self.node_tickets: Dict[kademlia.Node, NodeTicketInfo] = {}
+
+    def add_node(self, node: kademlia.Node, topic: bytes) -> None:
+        entries = self.topics.setdefault(topic, [])
```

Comment:
Did you consider using a `collections.defaultdict`?  That would allow you to remove this initialization step.

Reply:
I did that and got an error from mypy:

> No overload variant of "defaultdict" matches argument type "List[<nothing>]"

Any idea how to make that go away?

Reply:
A `cast` seems easiest.

```
self.topics = cast(Dict[bytes, List[TopicEntry]], defaultdict(list))
```

I haven't tried it but sometimes `mypy` knows how to handle the string versions:

```python
self.topics: "defaultdict[bytes, List[TopicEntry]]" = defaultdict(list)
```